### PR TITLE
Add uncertainty info into rule and evaluation

### PR DESCRIPTION
**Pull request by hwpang**

### CODE REVIEW

File: pysidt/sidt.py [hunk 1]
```diff
@@ -866,21 +891,36 @@ def estimate_uncertainty(self):
...
+            if n == 1:
+                node.rule.uncertainty = node.parent.rule.uncertainty
+
 
     def assign_depths(self):
         root = self.root
         _assign_depths(root)
 
-    def evaluate(self, mol):
+    def evaluate(self, mol, estimate_uncertainty=True):
```
Comment (mjohnson541) [inline pysidt/sidt.py]: default estimate_uncertainty=False for now

File: pysidt/sidt.py [hunk 2]
```diff
@@ -834,7 +859,7 @@ def fit_rule(self, alpha=0.1):
 
         logging.info("# nodes: {}".format(len(self.nodes)))
 
-    def estimate_uncertainty(self):
+    def estimate_uncertainty(self, confidence_level):
```
Comment (mjohnson541) [inline pysidt/sidt.py]: set default confidence_level 0.95?

File: pysidt/sidt.py [hunk 3]
```diff
@@ -9,10 +9,21 @@
 import json
 from sklearn import linear_model
 import scipy.sparse as sp
+import scipy
 
 logging.basicConfig(level=logging.INFO)
 
 
+class Rule:
```
Comment (mjohnson541) [inline pysidt/sidt.py:18]: This class won't json serialize so it should break write_nodes and read_nodes as currently written.

**Review by mjohnson541 — COMMENTED** [review]
A few comments as discussed offline.

File: pysidt/sidt.py (hunk 1 above)
Comment (hwpang) [reply pysidt/sidt.py]: Done

File: pysidt/sidt.py (hunk 2 above)
Comment (hwpang) [reply pysidt/sidt.py]: Done

File: pysidt/sidt.py (hunk 3 above)
Comment (hwpang) [reply pysidt/sidt.py:18]: I added a test to make sure that `write_nodes` and `read_nodes` work with `Rule`

**Comment by mjohnson541** [comment]
The remaining concern I have with this PR is that it fundamentally changes the file format for storing trees. I think this will make it impossible to load models generated with write_nodes before this PR. So we need to support loading of the old tree files that may assign an arbitrary object to the "rule" key rather than a Rule object.

**Comment by hwpang** [comment]
> The remaining concern I have with this PR is that it fundamentally changes the file format for storing trees. I think this will make it impossible to load models generated with write_nodes before this PR. So we need to support loading of the old tree files that may assign an arbitrary object to the "rule" key rather than a Rule object.

What if we support loading of the old tree files but turn the raw "rule" into a Rule object?

**Comment by hwpang** [comment]
I see that one of the jupyter notebook is failing the test with the following error traces. I will look into how to resolve this. I think it fails because of this line of change:

It was:

```
val == getattr(obj.__class__, attr):
```

I changed it to:

```
val == getattr(obj.__class__(), attr):
```

And the reason for this is that, I assume this line of code, it's trying to check that the val is the same as the default value assigned. I think the old version is wrong, because you need to instantiate to be able to have access to the attribute.

@mjohnson541 Do you have any idea how to resolve this? I would appreciate any advice.

Edit: I think I have resolved this successfully

```
=================================== FAILURES ===================================
_ /home/runner/work/PySIDT/PySIDT/IPython/Surface_Diffusion_single_eval_SIDT_example.ipynb _
---------------------------------------------------------------------------
file = "Surface_Diffusion_tree.json"
write_nodes(sidt, file)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
~/work/PySIDT/PySIDT/pysidt/sidt.py in write_nodes(tree, file)
    403         try:
--> 404             json.dumps(node.rule)
    405             rule = node.rule

~/miniconda3/envs/pysidt_env/lib/python3.7/json/__init__.py in dumps(obj, skipkeys, ensure_ascii, check_circular, allow_nan, cls, indent, separators, default, sort_keys, **kw)
    2[30](https://github.com/zadorlab/PySIDT/actions/runs/9454104801/job/26040928208?pr=18#step:6:31)         default is None and not sort_keys and not kw):
--> 2[31](https://github.com/zadorlab/PySIDT/actions/runs/9454104801/job/26040928208?pr=18#step:6:32)         return _default_encoder.encode(obj)
    232     if cls is None:

~/miniconda3/envs/pysidt_env/lib/python3.7/json/encoder.py in encode(self, o)
    198         # equivalent to the PySequence_Fast that ''.join() would do.
--> 199         chunks = self.iterencode(o, _one_shot=True)
    200         if not isinstance(chunks, (list, tuple)):

~/miniconda3/envs/pysidt_env/lib/python3.7/json/encoder.py in iterencode(self, o, _one_shot)
    256                 self.skipkeys, _one_shot)
--> 257         return _iterencode(o, 0)
    258 

~/miniconda3/envs/pysidt_env/lib/python3.7/json/encoder.py in default(self, o)
    178         """
--> 179         raise TypeError(f'Object of type {o.__class__.__name__} '
    180                         f'is not JSON serializable')

TypeError: Object of type SurfaceArrhenius is not JSON serializable

During handling of the above exception, another exception occurred:

TypeError                                 Traceback (most recent call last)
~/work/PySIDT/PySIDT/pysidt/sidt.py in to_dict(obj)
    351         try:
--> 352             json.dumps(val)
    353             out_dict[attr] = val

~/miniconda3/envs/pysidt_env/lib/python3.7/json/__init__.py in dumps(obj, skipkeys, ensure_ascii, check_circular, allow_nan, cls, indent, separators, default, sort_keys, **kw)
    230         default is None and not sort_keys and not kw):
--> 231         return _default_encoder.encode(obj)
    2[32](https://github.com/zadorlab/PySIDT/actions/runs/9454104801/job/26040928208?pr=18#step:6:33)     if cls is None:

~/miniconda3/envs/pysidt_env/lib/python3.7/json/encoder.py in encode(self, o)
    198         # equivalent to the PySequence_Fast that ''.join() would do.
--> 199         chunks = self.iterencode(o, _one_shot=True)
    200         if not isinstance(chunks, (list, tuple)):

~/miniconda3/envs/pysidt_env/lib/python3.7/json/encoder.py in iterencode(self, o, _one_shot)
    256                 self.skipkeys, _one_shot)
--> 257         return _iterencode(o, 0)
    258 

~/miniconda3/envs/pysidt_env/lib/python3.7/json/encoder.py in default(self, o)
    178         """
--> 179         raise TypeError(f'Object of type {o.__class__.__name__} '
    180                         f'is not JSON serializable')

TypeError: Object of type RateUncertainty is not JSON serializable

During handling of the above exception, another exception occurred:

TypeError                                 Traceback (most recent call last)
/tmp/ipykernel_2118/3947850130.py in <module>
      1 file = "Surface_Diffusion_tree.json"
----> 2 write_nodes(sidt, file)

~/work/PySIDT/PySIDT/pysidt/sidt.py in write_nodes(tree, file)
    406         except (TypeError, OverflowError):
    407             rule = to_dict(
--> 408                 node.rule
    409             )  # will work on all rmgmolecule objects, new objects need this method implemented
    410             try:

~/work/PySIDT/PySIDT/pysidt/sidt.py in to_dict(obj)
    362                 }
    363             else:
--> 364                 out_dict[attr] = to_dict(val)
    365 
    366     return out_dict

~/work/PySIDT/PySIDT/pysidt/sidt.py in to_dict(obj)
    [34](https://github.com/zadorlab/PySIDT/actions/runs/9454104801/job/26040928208?pr=18#step:6:35)6         val = getattr(obj, attr)
    347 
--> 348         if callable(val) or val == getattr(obj.__class__(), attr):
    349             continue
    [35](https://github.com/zadorlab/PySIDT/actions/runs/9454104801/job/26040928208?pr=18#step:6:36)0 

molecule/kinetics/uncertainties.pyx in molecule.kinetics.uncertainties.RateUncertainty.__init__()

TypeError: __init__() takes at least 3 positional arguments (0 given)
=========================== short test summary info ============================
FAILED IPython/Surface_Diffusion_single_eval_SIDT_example.ipynb:: - TypeError: __init__() takes at least 3 positional arguments (0 given)
============================== 1 failed in 3.59s ===============================
```

**Comment by mjohnson541** [comment]
> > The remaining concern I have with this PR is that it fundamentally changes the file format for storing trees. I think this will make it impossible to load models generated with write_nodes before this PR. So we need to support loading of the old tree files that may assign an arbitrary object to the "rule" key rather than a Rule object.
> 
> What if we support loading of the old tree files but turn the raw "rule" into a Rule object?

Yeah, that's what I was thinking of.

File: pysidt/sidt.py [hunk 4]
```diff
@@ -866,21 +907,36 @@ def estimate_uncertainty(self):
...
+        pred = 0.0
+        unc = 0.0
         decomp = self.decomposition(mol)
         for d in decomp:
             children = self.root.children
             node = self.root
-            out += node.rule
+            pred += node.rule.value
+            if estimate_uncertainty:
+                unc += node.rule.uncertainty ** 2
```
Comment (mjohnson541) [inline pysidt/sidt.py:941]: I was recently working with our uncertainties and I'm pretty sure our node uncertainties are in variance not standard deviation so I don't think this should be squared.

File: pysidt/sidt.py [hunk 5]
```diff
@@ -318,7 +346,7 @@ def to_dict(obj):
     for attr in attrs:
         val = getattr(obj, attr)
 
-        if callable(val) or val == getattr(obj.__class__, attr):
```
Comment (mjohnson541) [inline pysidt/sidt.py:321]: It seems like this is the change that's breaking the notebook. Why do we need it to be obj.__class__()?

File: pysidt/sidt.py (hunk 4 above)
Comment (hwpang) [reply pysidt/sidt.py:941]: <s>I added this line of code to calculate the uncertainty for rule, and didn't read it from the node_uncertainty dictionary:

```
                t = scipy.stats.t.ppf((1 + confidence_level) / 2, n - 1)
                node.rule = Rule(value=data_mean, uncertainty=t * np.std(data) / np.sqrt(n), num_data=n)
```

so the uncertainty under rule is the standard deviation. 

We should probably do the same for the node_uncertainty dict</s>

Edit:

Ok I checked. Please ignore above. For the single evaluation tree, I am using the standard deviation:

```
                t = scipy.stats.t.ppf((1 + confidence_level) / 2, n - 1)
                node.rule = Rule(value=data_mean, uncertainty=t * np.std(data) / np.sqrt(n), num_data=n)
```

For the multi evaluation tree, I am pulling from the node uncertainties dict:

```
                t = scipy.stats.t.ppf((1 + confidence_level) / 2, n - 1)
                node.rule.uncertainty = t * self.node_uncertainties[node.name] / np.sqrt(n)
```

I added a sqrt for the node uncertainties dict to make sure it's standard deviation:

```
            node_uncertainties = np.sqrt(
                np.diag(np.linalg.pinv((A.T @ A).toarray()))
                * (self.data_delta**2).sum()
                / (len(self.datums) - len(nodes))
            )
```

**Comment by hwpang** [comment]
> Yeah, that's what I was thinking of.

Ok done!

**Comment by hwpang** [comment]
@mjohnson541 Commits cleaned! This should be ready to merge.

**Review by mjohnson541 — APPROVED** [review]

**Pull request merged**